src/sentry/preprod/models.py

```diff
@@ -0,0 +1,209 @@
+            )
+
+    class ErrorCode:
+        UNKNOWN = 0
+        """The error code is unknown. Try to use a descriptive error code if possible."""
+        UPLOAD_TIMEOUT = 1
+        """The upload timed out."""
+        ARTIFACT_PROCESSING_TIMEOUT = 2
+        """The artifact processing timed out."""
+        ARTIFACT_PROCESSING_ERROR = 3
+        """The artifact processing failed."""
+
+        @classmethod
+        def as_choices(cls):
+            return (
+                (cls.UNKNOWN, "unknown"),
+                (cls.UPLOAD_TIMEOUT, "upload_timeout"),
+                (cls.ARTIFACT_PROCESSING_TIMEOUT, "artifact_processing_timeout"),
+                (cls.ARTIFACT_PROCESSING_ERROR, "artifact_processing_error"),
+            )
+
+    # Having a FK to both Org/Project is unnecessary
+    organization_id = BoundedBigIntegerField(db_index=True)
+    project = FlexibleForeignKey("sentry.Project")
+
+    # Nullable in case the file upload fails
+    file = FlexibleForeignKey("sentry.File", null=True)
+
+    # The date the artifact was built. E.g. an artifact could be built on 05/21/2025,
+    # but the user uploaded it on 05/22/2025.
+    date_built = models.DateTimeField(null=True)
+
+    build_configuration = models.ForeignKey("sentry.PreprodBuildConfiguration", null=True)
+
+    state = BoundedPositiveIntegerField(
+        default=ArtifactState.UPLOADING, choices=ArtifactState.as_choices()
+    )
+
+    artifact_type = BoundedPositiveIntegerField(choices=ArtifactType.as_choices(), null=True)
+
+    error_code = BoundedPositiveIntegerField(choices=ErrorCode.as_choices(), null=True)
+    error_message = models.TextField(null=True)
+
+    # E.g. 1.2.300
+    build_version = models.CharField(max_length=255, null=True)
+    # E.g. 9999
+    build_number = models.IntegerField(null=True)
+    # Some artifacts are embedded with a UUID, e.g. iOS Mach-O binaries
+    build_uuid = models.UUIDField(null=True)
+
+    misc = models.Field[dict[str, Any], dict[str, Any]] = JSONField(null=True)
+
+    class Meta:
+        app_label = "sentry"
+        db_table = "sentry_preprodartifact"
+
+
+class PreprodBuildConfiguration(DefaultFieldsModel):
+    """The build configuration used to build the artifact, e.g. "Debug" or "Release"."""
+
+    organization_id = BoundedBigIntegerField(db_index=True)
+    project = FlexibleForeignKey("sentry.Project")
+    name = models.CharField(max_length=255)
+
+    class Meta:
+        app_label = "sentry"
+        db_table = "sentry_preprodbuildconfiguration"
+        unique_together = ("project", "name")
+
+
+class PreprodArtifactSizeMetrics(DefaultFieldsModel):
+    """
+    Metrics about the size analysis of a pre-production artifact. Each PreprodArtifact can have 0 or many
+    size metrics.
+    """
+
+    class MetricsArtifactType:
+        MAIN_ARTIFACT = 0
+        """The main artifact."""
+        WATCH_ARTIFACT = 1
+        """An embedded watch artifact."""
+        ANDROID_DYNAMIC_FEATURE = 2
```

Comment:
Could add `ANDROID_INSTANT_APP`, but those are quite rare so can always add in the future if we need it

Reply:
Ah I thought that was considered a dynamic feature. I'll leave out for now since I'd likely have to add `IOS_APP_CLIP` too.